src/Database/LSMTree/Internal/BlobFile.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{- HLINT ignore "Use unless" -}
+
+module Database.LSMTree.Internal.BlobFile (
+    BlobFile (..)
+  , BlobSpan (..)
+  , removeReference
+  , newBlobFile
+  , readBlobFile
+  ) where
+
+import           Control.DeepSeq (NFData (..))
+import           Control.Monad.Class.MonadThrow (MonadThrow, MonadMask)
+import           Control.Monad.Primitive (PrimMonad)
+import           Control.RefCount (RefCounter)
+import qualified Control.RefCount as RC
+import qualified Data.Primitive.ByteArray as P (newPinnedByteArray,
+                     unsafeFreezeByteArray)
+import           Data.Word (Word32, Word64)
+import qualified Database.LSMTree.Internal.RawBytes as RB
+import           Database.LSMTree.Internal.Serialise (SerialisedBlob (..))
+import qualified System.FS.API as FS
+import           System.FS.API (HasFS)
+import qualified System.FS.BlockIO.API as FS
+
+-- | An open handle to a file containing blobs.
+--
+-- This is a reference counted object. Upon finalisation, the file is closed
+-- and deleted.
+--
+data BlobFile m h = BlobFile {
+       blobFileHandle     :: {-# UNPACK #-} !(FS.Handle h),
+       blobFileRefCounter :: {-# UNPACK #-} !(RefCounter m)
+     }
+  deriving stock (Show)
+
+instance NFData h => NFData (BlobFile m h) where
+  rnf (BlobFile a b) = rnf a `seq` rnf b
+
+-- | The location of a blob inside a blob file.
+data BlobSpan = BlobSpan {
+    blobSpanOffset :: {-# UNPACK #-} !Word64
+  , blobSpanSize   :: {-# UNPACK #-} !Word32
+  }
+  deriving stock (Show, Eq)
+
+instance NFData BlobSpan where
+  rnf (BlobSpan a b) = rnf a `seq` rnf b
+
+removeReference ::
+     (MonadMask m, PrimMonad m)
+  => BlobFile m h
+  -> m ()
+removeReference BlobFile{blobFileRefCounter} =
+    RC.removeReference blobFileRefCounter
+
+-- | Adopt an existing open file handle to make a 'BlobFile'. The file must at
+-- least be open for reading (but may or may not be open for writing).
+--
+-- The finaliser will close and delete the file.
+--
+newBlobFile ::
+     PrimMonad m
+  => HasFS m h
+  -> FS.Handle h
+  -> m (BlobFile m h)
+newBlobFile fs blobFileHandle = do
```

Comment:
I did it this way because the different use cases differ in how they handle the file: open for reading only in runs, but read/write in write buffer. I figured, just have it deal with closing. But yeah, we could change it, it'd need to take a read/write param.

Reply:
I've changed it to `openBlobFile`, taking the file path and open mode.

Reply:
The comment is a little outdated now

Reply:
Indeed. Fixed.